### PR TITLE
Issue: 716 - Refactor product service list and readone

### PR DESCRIPTION
**Pull request by zkhin**
Fixes: https://github.com/SeedCompany/cord-api-v3/issues/716

### CODE REVIEW

File: src/components/product/product.service.ts [hunk 1]
```diff
@@ -63,63 +74,116 @@ export class ProductService {
...
-      },
-    });
+      });
+
+      const products = await productQuery.run();
+
+      result.items = await Promise.all(
+        products.map(async (product) => this.readOne(product.id, session))
+      );
+      result.total = result.items.length;
```
Comment (michaelmarshall) [inline src/components/product/product.service.ts:162]: We don't want to run more than one DB query per function if we can help it.

File: src/components/product/product.service.ts [hunk 2]
```diff
@@ -63,63 +74,116 @@ export class ProductService {
...
+          { name: 'methodology', secure: false },
+        ],
+        input: {
+          page,
+          count,
+          sort,
+          order,
+          filter,
+        },
+      });
```
Comment (michaelmarshall) [inline src/components/product/product.service.ts:185]: We need to not use the old `database.service.ts` file functions. Instead, use the `query.helpers.ts` exports.

File: src/components/product/product.service.ts [hunk 3]
```diff
@@ -63,63 +74,116 @@ export class ProductService {
...
+            active: true,
+            id: $requestingUserId
+          }),
+          (engagement {id: $engagementId, active: true, owningOrgId: $owningOrgId})
+          ,(product:Product {active:true})
+        WITH COUNT(product) as total, engagement, product
+          RETURN total, product.id as id, product.createdAt as createdAt
+          ORDER BY ${sort} ${order}
+          SKIP $skip LIMIT $count
+      `;
```
Comment (michaelmarshall) [inline src/components/product/product.service.ts:147]: We need to compose the query using the functions from `query.helpers.ts`. The `organization` service's `list` function is a good example.

File: src/components/product/product.service.ts [hunk 4]
```diff
@@ -63,63 +74,116 @@ export class ProductService {
...
-      approach: result.approach.value,
-      methodology: result.methodology.value,
+      createdAt: result.createdAt,
+      type: result.type,
+      books: result.books,
+      mediums: result.mediums,
+      purposes: result.purposes,
+      approach: result.approach,
+      methodology: result.methodology,
     };
```
Comment (michaelmarshall) [inline src/components/product/product.service.ts:118]: This needs to be in the `query.helpers.ts` style. Use the `organization` service's `readOne` function as an example.

**Comment by zkhin** [comment]
@michaelmarshall I think I need to implement `filterByEngagement` in `query.helper.ts` for filtering products by engagementId. Maybe it's similar to `filterByUser` but let me that I can use `:engagement` relationshipType.

File: src/components/product/product.service.ts [hunk 5]
```diff
@@ -23,106 +37,141 @@ import {
...
+    const props = ['mediums', 'purposes', 'methodology'];
+    const query = this.db
+      .query()
+      .call(matchRequestingUser, session)
+      .call(matchUserPermissions, 'Product', id)
+      .call(addAllSecureProperties, ...props)
+      .with([
+        ...props.map(addPropertyCoalesceWithClause),
+        'coalesce(node.id) as id',
+        'coalesce(node.createdAt) as createdAt',
```
Comment (CarsonF) [inline src/components/product/product.service.ts:163]: FYI there's no need to `coalesce` if there's not more than one argument in the function call.

File: src/core/database/query.helpers.ts [hunk 6]
```diff
@@ -621,6 +621,22 @@ export function filterByUser(
   ]);
 }
 
+// used to search a specific engagement's relationship to the target base node
+// for example, searching all products a engagement is a part of
+export function filterByEngagement(
```
Comment (CarsonF) [inline src/core/database/query.helpers.ts:626]: This is specific to engagements. It should live in the engagements folder not core.

File: src/components/product/product.service.ts [hunk 7]
```diff
@@ -242,4 +216,109 @@ export class ProductService {
...
+    const query = this.db
+      .query()
+      .call(matchRequestingUser, session)
+      .call(matchUserPermissions, 'Product');
+
+    if (filter.methodology) {
+      query.call(filterByString, label, 'methodology', filter.methodology);
+    } else if (filter.approach) {
+      query.call(filterByString, label, 'approach', filter.approach);
+    } else if (filter.engagementId) {
```
Comment (CarsonF) [inline src/components/product/product.service.ts:238]: These filters are not either/or. They all could be applied at once.

File: src/components/product/product.service.ts [hunk 8]
```diff
@@ -23,106 +37,141 @@ import {
...
+    if (!result) {
+      throw new ServerException('failed to create default product');
     }
 
+    const id = result.id;
+
+    // add root admin to new product as an admin
+    await this.db.addRootAdminToBaseNodeAsAdmin(id, 'Product');
+
+    this.logger.debug(`product created, id ${id}`);
```
Comment (CarsonF) [inline src/components/product/product.service.ts:108]:
```suggestion
    this.logger.debug(`product created`, { id });
```

**Pull request merged**